src/org/mozilla/javascript/SpecialRef.java

```diff
@@ -105,11 +105,13 @@ public Object set(Context cx, Scriptable scope, Object value)
                     }
 
                     if (cx.getLanguageVersion() >= Context.VERSION_ES6) {
-                        if ((value != null && !"object".equals(ScriptRuntime.typeof(value))) ||
-                            !"object".equals(ScriptRuntime.typeof(target))) {
-                            return Undefined.instance;
+                        if (value != null && !"object".equals(ScriptRuntime.typeof(value))) {
```

Comment:
```suggestion
                        final String typeOfValue = ScriptRuntime.typeof(value);
                        if (!"object".equals(typeOfValue) && !"function".equals(typeOfValue)) {
```

Reply:
> 2. If Type(proto) is neither Object nor Null, return undefined.

Since it is written like this, I think it is easier to see that `value != null` is more in line with the specification.
If you want to remove this, I would like the following test case.
```js
var a = {};
a.__proto__ = null;
a.__proto__ === undefined; // true
```